primitives/account/tests/staking_contract/staker.rs

```diff
@@ -346,388 +425,505 @@ fn stake_works() {
     assert_eq!(
         staker_setup
             .staking_contract
             .active_validators
-            .get(&validator_address2),
-        Some(&Coin::from_u64_unchecked(Policy::VALIDATOR_DEPOSIT))
+            .get(&validator_address),
+        Some(&validator_deposit)
     );
+}
 
-    // Doesn't work when the staker doesn't exist.
-    let fake_keypair = ed25519_key_pair(VALIDATOR_PRIVATE_KEY);
-
-    let tx = make_signed_incoming_transaction(
-        IncomingStakingTransactionData::UpdateStaker {
-            new_delegation: None,
-            reactivate_all_stake: false,
-            proof: SignatureProof::default(),
-        },
+#[test]
+fn cannot_retire_active_stake() {
+    // -----------------------------------
+    // Test setup:
+    // -----------------------------------
+    let mut staker_setup = StakerSetup::setup_staker_with_inactive_retired_balance(
+        ValidatorState::Active,
+        Policy::MINIMUM_STAKE,
         0,
-        &fake_keypair,
+        1,
     );
+    let data_store = staker_setup
+        .accounts
+        .data_store(&Policy::STAKING_CONTRACT_ADDRESS);
+    let mut db_txn = staker_setup.env.write_transaction();
+    let mut db_txn = (&mut db_txn).into();
 
+    // -----------------------------------
+    // Test execution:
+    // -----------------------------------
+    // Cannot retire funds from active.
+    let tx = make_retire_stake_transaction(Policy::MINIMUM_STAKE);
+    let block_state = BlockState::new(
+        Policy::block_after_reporting_window(Policy::election_block_after(2)),
+        2,
+    );
+    let mut tx_logger = TransactionLog::empty();
     assert_eq!(
         staker_setup.staking_contract.commit_incoming_transaction(
             &tx,
             &block_state,
             data_store.write(&mut db_txn),
-            &mut TransactionLog::empty()
+            &mut tx_logger,
         ),
-        Err(AccountError::NonExistentAddress {
-            address: (&fake_keypair.public).into()
+        Err(AccountError::InsufficientFunds {
+            needed: Coin::from_u64_unchecked(Policy::MINIMUM_STAKE),
+            balance: Coin::ZERO
         })
     );
 }
 
 #[test]
-fn update_staker_with_stake_reactivation_works() {
+fn update_staker_works() {
     // -----------------------------------
     // Test setup:
     // -----------------------------------
-    let mut staker_setup =
-        StakerSetup::setup_staker_with_inactive_balance(ValidatorState::Active, 0, 150_000_000);
+    let (mut staker_setup, validator_address2, tx) =
+        prepare_second_validator_for_redelegation(ValidatorState::Active, 0, 150_000_000, 0);
+
     let data_store = staker_setup
         .accounts
         .data_store(&Policy::STAKING_CONTRACT_ADDRESS);
     let mut db_txn = staker_setup.env.write_transaction();
     let mut db_txn = (&mut db_txn).into();
-    let mut data_store_write = data_store.write(&mut db_txn);
+    _ = data_store.write(&mut db_txn);
```

Comment:
What does this do? (Also the further occurrences in this file.)

Reply:
This is a bit annoying. So without that write, the .into() from the line above infers the wrong type because the first operation performed in these tests using the `db_txn` is a `get_staker`. This is a read operation on the datastore and thus when we use it to write (on the commit incoming transaction) we have a type mismatch. 
This is totally a hardcoded way of dealing with the problem, let me know if there is a cleaner way.

Reply:
I guess a type annotation on the second `db_txn` would work?

```rs
use nimiq_trie::WriteTransactionProxy;

let mut db_txn = staker_setup.env.write_transaction();
let mut db_txn: WriteTransactionProxy = (&mut db_txn).into();
```

Reply:
This was so damn stupid xD I had tried it but rust analyzer was complaining still, I probably had the wrong import. 🙈